### PR TITLE
Navigation screen: Add opt-in Navigation block rendering

### PR DESCRIPTION
**Pull request by noisysocks**
Implements the majority of https://github.com/WordPress/gutenberg/issues/24364. Continues on from the work started in https://github.com/WordPress/gutenberg/pull/22656.

demo

Allows themes to opt-in to having `wp_nav_menu()` render a complete Navigation block in place of the menu. This allows arbitrarily complex Navigation block structures can be used in an existing theme.
    
When a theme declares `add_theme_support( 'block-nav-menus' )`, users are able to add non-link blocks to a Navigation using the Navigation screen.
    
This block tree is stored as menu items and re-assembled into a Navigation block for display using `render_block()` in the frontend.

If the user switches to a theme that does not support block menus, or disables this functionality, non-link blocks are no longer rendered in the frontend. Care is taken, however, to ensure that users can still see their data in `nav-menus.php`.

See https://github.com/WordPress/gutenberg/issues/24364 for more information on the approach.

### How to test

1. Note that in _Gutenberg > Navigation (beta)_, you can only add Link blocks to your Navigation. Note also that the frontend renders regular nav menu markup.

2. Patch your current theme to support block navigation. Here's how that looks for Twenty Twenty:
   
```diff
Index: src/wp-content/themes/twentytwenty/functions.php
===================================================================
--- src/wp-content/themes/twentytwenty/functions.php	(revision 48655)
+++ src/wp-content/themes/twentytwenty/functions.php	(working copy)
@@ -142,6 +142,7 @@
 	$loader = new TwentyTwenty_Script_Loader();
 	add_filter( 'script_loader_tag', array( $loader, 'filter_script_loader_tag' ), 10, 2 );
 
+	add_theme_support( 'block-nav-menus' );
 }
 
 add_action( 'after_setup_theme', 'twentytwenty_theme_support' );
Index: src/wp-content/themes/twentytwenty/header.php
===================================================================
--- src/wp-content/themes/twentytwenty/header.php	(revision 48655)
+++ src/wp-content/themes/twentytwenty/header.php	(working copy)
@@ -83,25 +83,25 @@
 
 					<?php
 					if ( has_nav_menu( 'primary' ) || ! has_nav_menu( 'expanded' ) ) {
-						?>
 
-							<nav class="primary-menu-wrapper" aria-label="<?php esc_attr_e( 'Horizontal', 'twentytwenty' ); ?>" role="navigation">
+						if ( has_nav_menu( 'primary' ) ) {
 
-								<ul class="primary-menu reset-list-style">
+							wp_nav_menu(
+								array(
+									'container'      => '',
+									'items_wrap'     => '%3$s',
+									'theme_location' => 'primary',
+								)
+							);
 
-								<?php
-								if ( has_nav_menu( 'primary' ) ) {
+						} elseif ( ! has_nav_menu( 'expanded' ) ) {
+							?>
 
-									wp_nav_menu(
-										array(
-											'container'  => '',
-											'items_wrap' => '%3$s',
-											'theme_location' => 'primary',
-										)
-									);
+							<nav class="primary-menu-wrapper" aria-label="<?php esc_attr_e( 'Horizontal', 'twentytwenty' ); ?>" role="navigation">
 
-								} elseif ( ! has_nav_menu( 'expanded' ) ) {
+								<ul class="primary-menu reset-list-style">
 
+									<?php
 									wp_list_pages(
 										array(
 											'match_menu_classes' => true,
@@ -110,15 +110,14 @@
 											'walker'   => new TwentyTwenty_Walker_Page(),
 										)
 									);
+									?>
 
-								}
-								?>
-
 								</ul>
 
 							</nav><!-- .primary-menu-wrapper -->
 
-						<?php
+							<?php
+						}
 					}
 
 					if ( true === $enable_header_search || has_nav_menu( 'expanded' ) ) {
```

2. The frontend should now render Navigation block markup.

3. Go back to _Gutenberg > Navigation (beta)_ and add a Search block. The Search block should now display on the frontend.

4. Switch to a different theme. The frontend should now display regular nav menu markup without a Search block.

5. Go back to _Gutenberg > Navigation (beta)_. The editor should now not allow a Search block to be inserted. The existing Search block should still be displayed.

6. Go to _Appearance > Menus_. A Gutenberg Block widget should appear containing the Search block.

### Follow up tasks

- Investigate making this setting _per menu location_. https://github.com/WordPress/gutenberg/pull/24503#issuecomment-673198300
- Add UI which indicates when a block is not going to be rendered on the frontend. https://github.com/WordPress/gutenberg/pull/24503#discussion_r471872542
- Investigate registering a different 'missing' block. https://github.com/WordPress/gutenberg/pull/24503#discussion_r471894496

### CODE REVIEW

**Comment by manmotive** [comment]
@noisysocks I'd like to see a way here for wp_nav_menu to be used in conjunction with blocks, and for plugins to be able to switch it on block support without affecting other locations in the theme.

I don't know the exact answer, but I think there needs to be further thought on how adding theme_support('block-nav-menus') applies to each instance of wp_nav_menu. Turning it on would mean the theme author would need to update _every_ menu on their theme, but they won't be able to account for custom menus that their users have added, and it would wipe out any user modifications made using filters/actions/plugins that only walkers support (on that note, there would need to be some mechanism for the user to opt out of the block html navigation and go back to the walker). And the other way around, if a plugin added theme_support('block-nav-menus'), it would apply to _all_ menu locations in the theme.

In short I think a more granular approach is needed:
1. add_theme_support( 'block-nav-menus' ) enables blocks to be added to the nav-menus.php page
2. a wp_nav_menu argument needs to be turned on to allow blocks to be output or to use the gutenberg block (additionally, wp_nav_menus using a custom walker should never use the block html output). E.g.
$args->block_support = false (walker without blocks) / walker (walker with blocks) / true (use navigation block html)

(Terrible naming but you get the idea!)

Allowing 3 situations to be accounted for:
1. A walker menu without blocks
2. A walker menu with blocks
3. A navigation block

(all 3 may be used in conjunction on the same site/theme)

**Comment by noisysocks** [comment]
Hey @manmotive, thanks for the feedback. Making this setting more granular is a good idea. How about we make it a _per menu location_ setting instead of a _per theme_ setting by updating [`register_nav_menu`](https://developer.wordpress.org/reference/functions/register_nav_menu) to accept a `supports` argument?

```php
register_nav_menu(
	'primary',
	array(
		'description' => __( 'Desktop Horizontal Menu' ),
		'supports'    => array(
			'blocks' => true,
		),
	)
);
```

It's a straightforward change but it will require Core changes because `register_nav_menu` and `register_nav_menus` are not filterable. For this reason, let's tackle it separately to this PR.

**Comment by noisysocks** [comment]
This is ready for review. I updated the PR description with a demo GIF and testing steps.

This now supports storing arbitrarily complex block trees which you can test for yourself by adding e.g. `core/group` to the list of allowed blocks in `packages/block-library/src/navigation/edit.js`. This gives us a great deal of flexibility for future changes to the Navigation block. For example, we can have container blocks in Navigation and still use them in existing themes.

Here's a very contrived example of how a `Group > Columns > Column (x3) > Image | Link` structure looks when viewed in `nav-menus.php`. Note that the deeply nested Link blocks are still converted to a menu item that older themes can understand.

Screen Shot 2020-08-13 at 14 44 19

File: lib/navigation.php [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+			}
+
+			.menu-item-block textarea {
+				width: 100%;
+				min-height: 100px;
+			}
+CSS;
+		wp_add_inline_style( 'nav-menus', $css );
+	}
+}
```
Comment (talldan) [inline lib/navigation.php:354]: This doesn't take into account that the user might have other fields enabled in screen options.

Might be best to add a unique class to the custom field wrapper, and then use `:not` to hide all the others.

Having said that, I wonder if we can get away with just CSS hacks. An issue I spotted is that it's possible to nest blocks in the nav-menus screen in a way that's incorrect (e.g. adding a link to search) ...

File: packages/edit-navigation/src/index.js [hunk 2]
```diff
@@ -16,13 +16,21 @@ import { __ } from '@wordpress/i18n';
...
 import Layout from './components/layout';
 import './store';
 
+function disableInsertingNonNavigationBlocks( settings, name ) {
+	if ( ! [ 'core/navigation', 'core/navigation-link' ].includes( name ) ) {
+		set( settings, [ 'supports', 'inserter' ], false );
+	}
+	return settings;
+}
+
```
Comment (talldan) [inline packages/edit-navigation/src/index.js:33]: Could also just register the blocks we want instead of registering every core and experimental block and then disabling inserter support.

File: lib/navigation.php [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+ * This shim can be removed when the Gutenberg plugin requires a WordPress
+ * version that has the ticket below.
+ *
+ * @see https://core.trac.wordpress.org/ticket/50544
+ *
+ * @param object $menu_item The menu item object.
+ */
+function gutenberg_setup_block_nav_menu_item( $menu_item ) {
+	if ( 'block' === $menu_item->type ) {
+		$menu_item->type_label = __( 'Gutenberg Block', 'gutenberg' );
```
[the remark this answers is not in the document]
Comment (draganescu) [reply lib/navigation.php:76]: I second the point to remove the word "Gutenberg" :) Just "Block", or yes, ideally, the block's name.

File: lib/navigation.php (hunk 1 above)
Comment (draganescu) [reply lib/navigation.php:354]: >  it's possible to nest blocks in the nav-menus screen in a way that's incorrect (e.g. adding a link to search) .
That is something. If they're menu items I am not sure we can make them un-nestable.

**Review by draganescu — COMMENTED** [review]
I have demoed this twice by now and it works quite (amazingly) well! Aside from @talldan 's review bits, it is a solid next step and quite merge-able.

File: lib/navigation.php [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+				$block = $parsed_blocks[0];
+			} else {
+				$block = array(
+					'blockName' => 'core/freeform',
+					'attrs'     => array(
+						'originalContent' => $menu_item->content,
+					),
+				);
+			}
+		} else {
```
[the remark this answers is not in the document]
Comment (noisysocks) [reply lib/navigation.php:184]: I think it's OK. The type is either going to be one of the Core types (`custom`, `object`, etc.) or a totally custom one that a plugin adds. Either way we should use `title` and `url` to create a link. At least until https://github.com/WordPress/gutenberg/issues/22919, that is.

File: lib/navigation.php [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+			$block = array(
+				'blockName' => 'core/navigation-link',
+				'attrs'     => array(
+					'label' => $menu_item->title,
+					'url'   => $menu_item->url,
+				),
+			);
```
[the remark this answers is not in the document]
Comment (noisysocks) [reply lib/navigation.php:191]: Yes 100%!

File: lib/navigation.php (hunk 1 above)
Comment (noisysocks) [reply lib/navigation.php:354]: > Might be best to add a unique class to the custom field wrapper, and then use :not to hide all the others.

Done in 641aebb17b. It relies on extra fields having the `.description` class which isn't necessarily the case for e.g. third party plugins. I think it's the best we can do without patching Core.

> Having said that, I wonder if we can get away with just CSS hacks. An issue I spotted is that it's possible to nest blocks in the nav-menus screen in a way that's incorrect (e.g. adding a link to search) ...

That's an interesting point... maybe we could disable re-ordering Block menu items? I'm not sure.

File: packages/edit-navigation/src/index.js (hunk 2 above)
Comment (noisysocks) [reply packages/edit-navigation/src/index.js:33]: The idea is to have them registered but not insertable so that, when a user switches away from a theme that supports blocks, they'll still see the blocks that they previously added to their menu. This way we're not "losing" any of the user's data. 

We'll need to add some sort of UI which indicates that the invalid blocks won't display in the frontend. I think best to do that separately since it requires design.

File: lib/navigation.php [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+ * @see https://core.trac.wordpress.org/ticket/50544
+ *
+ * @param int     $item_id Menu item ID.
+ * @param WP_Post $item    Menu item data object.
+ */
+function gutenberg_output_block_menu_item_custom_fields( $item_id, $item ) {
+	if ( 'block' === $item->type ) {
+		?>
+		<p class="description description-wide">
+			<textarea readonly><?php echo esc_textarea( trim( $item->content ) ); ?></textarea>
```
[the remark this answers is not in the document]
Comment (noisysocks) [reply lib/navigation.php]: c6b0a8813b

File: lib/navigation.php [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+ * Specifically, this shim makes it so that `wp_nav_menu()` will remove any menu
+ * items that have a `type` of `'block'` from `$sorted_menu_items`. When merged
+ * to Core, this functionality should exist in `wp_nav_menu()`.
+ *
+ * This shim can be removed when the Gutenberg plugin requires a WordPress
+ * version that has the ticket below.
+ *
+ * @see https://core.trac.wordpress.org/ticket/50544
+ *
+ * @param array $menu_items The menu items, sorted by each menu item's menu order.
```
[the remark this answers is not in the document]
Comment (noisysocks) [reply lib/navigation.php:137]: 38da717124

File: lib/navigation.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (noisysocks) [reply lib/navigation.php]: > It'd be awesome to have the block name, but that'd probably be hard work.

We could display the block's `name` (e.g. `"core/image"`) but I'm not sure that that's very useful to users.

I changed it to 'Block' in 50d192fff2. I updated the widgets screen, too, for consistency.

File: lib/navigation.php (hunk 1 above)
Comment (talldan) [reply lib/navigation.php]: I wonder if we should not allow using this UI at all if the menu contains blocks.

Is there a purpose to have multiple interfaces for the same thing?

File: packages/edit-navigation/src/index.js (hunk 2 above)
Comment (talldan) [reply packages/edit-navigation/src/index.js:33]: I see what you mean, though not registering the blocks seems fairly close to what you describe. If they're not registered, we wouldn't lose the data, but the block would be shown as unrecognized/missing.

For this page we could also consider registering a different 'missing' block that handles the messaging better:
https://github.com/WordPress/gutenberg/tree/master/packages/block-library/src/missing

File: lib/navigation.php (hunk 1 above)
Comment (noisysocks) [reply lib/navigation.php]: We'll need to support disabling the new screen and functionality entirely (e.g. using the Classic Editor plugin) because there will be some users that prefer the old screen and some setups that rely on the `nav-menus.php` DOM structure. 

If the new screen and functionality is disabled, I don't want any blocks that were inserted into a menu to go "missing". Hence the need to display them in `nav-menus.php`.

File: packages/edit-navigation/src/index.js (hunk 2 above)
Comment (noisysocks) [reply packages/edit-navigation/src/index.js:33]: > For this page we could also consider registering a different 'missing' block that handles the messaging better:
https://github.com/WordPress/gutenberg/tree/master/packages/block-library/src/missing

Good idea!

> Perhaps move this particular part out into a separate issue.

Agreed.

File: lib/navigation.php [hunk 8]
```diff
@@ -323,7 +336,7 @@ function gutenberg_output_block_menu_item_custom_fields( $item_id, $item ) {
 		<p class="field-content description description-wide">
 			<label for="edit-menu-item-content-<?php echo $item_id; ?>">
 				<?php _e( 'Content', 'gutenberg' ); ?><br />
-				<textarea readonly><?php echo esc_textarea( trim( $item->content ) ); ?></textarea>
+				<textarea readonly name="menu-item-content[<?php echo $item_id; ?>]"><?php echo esc_textarea( trim( $item->content ) ); ?></textarea>
```
Comment (talldan) [inline lib/navigation.php:339]: Seeing some square brackets in the in the interpolation for the `name` here:
Screenshot 2020-08-18 at 5 39 26 pm

**Review by talldan — APPROVED** [review]
Looks good to me, nice work here! Just fix that issue with the `name` attribute before merging. 🚀 

I think there are a couple of follow-ups:
- Decide how to deal with non-navigation blocks on the old nav screen (e.g. disallow moving them to prevent invalid nesting?)
- Decide how to deal with non-navigation blocks on the new nav screen for themes that opt out of the feature.

File: lib/navigation.php (hunk 8 above)
Comment (noisysocks) [reply lib/navigation.php:339]: That's correct. PHP interprets form values with `[]` as an array.

https://www.php.net/manual/en/reserved.variables.post.php#87650

**Comment by noisysocks** [comment]
Thanks all! I created https://github.com/WordPress/gutenberg/issues/24638, https://github.com/WordPress/gutenberg/issues/24639 and https://github.com/WordPress/gutenberg/issues/24640 as follow up issues.
